File: Leetcode/121-Best-Time-to-Buy-and-Sell-Stock.js
/**
 * @param {number[]} prices
 * @return {number}
 * // TC: O(N) SC: O(1)
 */
 var maxProfit = function(prices) {
    // I buy low sell high
    let left = 0;
    let diff = 0;
    let max = 0;
    for(let right = 1; right < prices.length; right++){
        if (prices[right] <= prices[left]){
            left = right
        }else {
            diff = prices[right] - prices[left];
            max = Math.max(diff, max)
        }

    }
    return max;
};

/**
 * @param {number[]} prices
 * @return {number}
 * // TC: O(N) SC: O(1)
 */
 var maxProfit = function(arr) {
    let profit = 0;
    let lowest = arr[0];
    for (let i=1;i < arr.length; i++) {
        if (arr[i] < lowest) lowest = arr[i];
        profit = Math.max(profit, arr[i] - lowest) 
    } 
 
    return profit;
};


// console.log(maxProfit([7,1,5,3,6,4]));
console.log(maxProfit([7,6,4,3,1]));

File: Leetcode/121. Best Time to Buy and Sell Stock.py
# Brute force TC: O(N^2) SC: O(1)

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        maxP = 0
        for i in range(len(prices)):
            for j in range(i + 1, len(prices)):
                if prices[j] > prices[i]:
                    maxP = max(maxP, prices[j] - prices[i])
        
        return maxP

# Two Pointer TC: O(N) SC: O(1)
        
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        maxProfit = 0
        buy = 0

        for sell in range(1,len(prices)):
            # when do i sell
            # only when buy price is lesser than sell price
            # I buy whenever I encounter a price cheaper than my previous buying price
            # buy low sell high
            # [8, 3, 6, 5, 1, 7]

            if prices[buy] > prices[sell]:
                buy = sell

            else: 
                profit = prices[sell] - prices[buy]
                maxProfit = max(profit, maxProfit)

        return maxProfit